{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "plink2 = '/home/vitaled2/.genotools/misc/executables/plink2'\n",
    "project_dir = '/home/vitaled2/gp2_carriers'\n",
    "data_dir = f'{project_dir}/data'\n",
    "report_path = f'{data_dir}/NBA_Report.csv'\n",
    "geno_dir = f'{data_dir}/raw_genotypes'\n",
    "labels = ['AAC', 'AFR', 'AJ', 'AMR', 'CAH', 'CAS', 'EAS', 'EUR', 'FIN','MDE', 'SAS']\n",
    "key = pd.read_csv(f'data/nba_app_key.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get snps list\n",
    "report = pd.read_csv(report_path)\n",
    "report.columns = [x.strip() for x in report.columns]\n",
    "report.drop(columns=['GP2ID','Cohort Code', 'Local ID',\t'PI','PD EUR PRS (GWAS vXX)','NBA derived Ancestry'], inplace=True)\n",
    "\n",
    "snp_df = report.loc[~report['Unnamed: 0'].isna()]\n",
    "snp_df.set_index('Unnamed: 0', inplace=True)\n",
    "snp_df = snp_df.transpose()\n",
    "snp_df.reset_index(inplace=True)\n",
    "snp_df.rename(columns={'index':'snp'}, inplace=True)\n",
    "snp_df.drop(columns=['Alternative variant name'], inplace=True)\n",
    "snp_df.columns.name = None\n",
    "snp_df.columns = ['snp','rsid','hg19','hg38','ancestry']\n",
    "snp_df = snp_df.loc[~snp_df.hg38.isna()]\n",
    "\n",
    "pvar_total = pd.DataFrame()\n",
    "for label in labels:\n",
    "    label_geno_path = f'{geno_dir}/{label}/{label}_release9_vwb'\n",
    "    pvar = pd.read_csv(f'{label_geno_path}.pvar', sep='\\t', dtype={'#CHROM':str})\n",
    "    pvar.columns = ['chrom', 'pos', 'id', 'a1', 'a2']\n",
    "    pvar.loc[:,'snpid1'] = pvar['chrom'] + ':' + pvar['pos'].astype(str) + ':' + pvar['a1'] + ':' + pvar['a2']\n",
    "    pvar.loc[:,'snpid2'] = pvar['chrom'] + ':' + pvar['pos'].astype(str) + ':' + pvar['a2'] + ':' + pvar['a1']\n",
    "    pvar_sub1 = pvar.loc[pvar['snpid1'].isin(snp_df['hg38'])].copy()\n",
    "    pvar_sub1.loc[:,'snpid'] = pvar_sub1['snpid1']\n",
    "    pvar_sub2 = pvar.loc[pvar['snpid2'].isin(snp_df['hg38'])].copy()\n",
    "    pvar_sub2.loc[:,'snpid'] = pvar_sub2['snpid2']\n",
    "    pvar_total = pd.concat([pvar_total, pvar_sub1, pvar_sub2], ignore_index=True)\n",
    "pvar_total.drop(columns=['snpid1','snpid2'], inplace=True)\n",
    "\n",
    "# pvar_total[['id']].drop_duplicates().to_csv(f'{data_dir}/snps.txt', header=False, index=False)\n",
    "\n",
    "snp_df_out = pvar_total.merge(snp_df, how='left', left_on='snpid', right_on='hg38')\n",
    "snp_df_out.drop(columns=['chrom', 'pos', 'a1', 'a2'], inplace=True)\n",
    "snp_df_out.rename(columns={'snp':'variant'}, inplace=True)\n",
    "\n",
    "snp_df_out['snp_name'] = snp_df_out['variant'].str.extract(f'\\((.*?)\\)')\n",
    "\n",
    "snp_df_out.loc[:,'locus'] = snp_df_out.variant.str.split(' ', expand=True)[:][0]\n",
    "snp_df_out.loc[:,'snp_name_full'] = snp_df_out['locus'] + '_' + snp_df_out['snp_name'] + '_' + snp_df_out['id']\n",
    "snp_df_out.drop(columns=['snpid'], inplace=True)\n",
    "snp_df_out.drop_duplicates(inplace=True)\n",
    "snp_df_out.to_csv(f'{data_dir}/snps_out.csv', index=False)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/AAC/AAC_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/AAC/AAC_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/AAC/AAC_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/AAC/AAC_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:48:04 2025\n",
      "32105 MiB RAM detected, ~30169 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "1230 samples (724 females, 506 males; 1230 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AAC/AAC_release9_vwb.psam.\n",
      "1896835 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AAC/AAC_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (341 cases, 837 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AAC/AAC_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AAC/AAC_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:48:04 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/AFR/AFR_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/AFR/AFR_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/AFR/AFR_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/AFR/AFR_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:48:04 2025\n",
      "32105 MiB RAM detected, ~30178 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "2767 samples (1244 females, 1523 males; 2767 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AFR/AFR_release9_vwb.psam.\n",
      "1788859 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AFR/AFR_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (992 cases, 1730 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AFR/AFR_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AFR/AFR_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:48:05 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/AJ/AJ_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/AJ/AJ_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/AJ/AJ_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/AJ/AJ_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:48:05 2025\n",
      "32105 MiB RAM detected, ~30167 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "3251 samples (1213 females, 2038 males; 3251 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AJ/AJ_release9_vwb.psam.\n",
      "1875579 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AJ/AJ_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1497 cases, 431 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AJ/AJ_release9_vwb_snps.afreq .\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AJ/AJ_release9_vwb_snps.traw ...\n",
      "121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:48:06 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/AMR/AMR_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/AMR/AMR_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/AMR/AMR_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/AMR/AMR_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:48:07 2025\n",
      "32105 MiB RAM detected, ~30164 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "3581 samples (1879 females, 1702 males; 3581 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AMR/AMR_release9_vwb.psam.\n",
      "1843508 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AMR/AMR_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1972 cases, 1457 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AMR/AMR_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/AMR/AMR_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:48:07 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/CAH/CAH_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/CAH/CAH_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/CAH/CAH_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/CAH/CAH_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:48:08 2025\n",
      "32105 MiB RAM detected, ~30168 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "1034 samples (493 females, 541 males; 1034 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/CAH/CAH_release9_vwb.psam.\n",
      "1904683 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/CAH/CAH_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (668 cases, 322 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/CAH/CAH_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/CAH/CAH_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:48:08 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/CAS/CAS_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/CAS/CAS_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/CAS/CAS_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/CAS/CAS_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:48:09 2025\n",
      "32105 MiB RAM detected, ~30158 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "1089 samples (578 females, 511 males; 1089 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/CAS/CAS_release9_vwb.psam.\n",
      "1902516 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/CAS/CAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (617 cases, 347 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/CAS/CAS_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/CAS/CAS_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:48:09 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/EAS/EAS_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/EAS/EAS_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/EAS/EAS_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/EAS/EAS_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:48:09 2025\n",
      "32105 MiB RAM detected, ~30159 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "5902 samples (2251 females, 3651 males; 5902 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/EAS/EAS_release9_vwb.psam.\n",
      "1730431 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/EAS/EAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (1882 cases, 2445 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/EAS/EAS_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/EAS/EAS_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:48:09 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/EUR/EUR_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/EUR/EUR_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/EUR/EUR_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/EUR/EUR_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:48:11 2025\n",
      "32105 MiB RAM detected, ~30152 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "51247 samples (22571 females, 28676 males; 51247 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/EUR/EUR_release9_vwb.psam.\n",
      "1668646 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/EUR/EUR_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (23198 cases, 10230 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/EUR/EUR_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/EUR/EUR_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:48:12 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/FIN/FIN_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/FIN/FIN_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/FIN/FIN_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/FIN/FIN_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:49:12 2025\n",
      "32105 MiB RAM detected, ~30021 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "137 samples (78 females, 59 males; 137 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/FIN/FIN_release9_vwb.psam.\n",
      "1883041 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/FIN/FIN_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (106 cases, 9 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/FIN/FIN_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/FIN/FIN_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:49:12 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/MDE/MDE_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/MDE/MDE_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/MDE/MDE_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/MDE/MDE_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:49:12 2025\n",
      "32105 MiB RAM detected, ~30025 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "866 samples (336 females, 530 males; 866 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/MDE/MDE_release9_vwb.psam.\n",
      "1888776 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/MDE/MDE_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (477 cases, 224 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/MDE/MDE_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/MDE/MDE_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:49:12 2025\n",
      "PLINK v2.0.0-a.6.3LM AVX2 Intel (3 Dec 2024)       cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/vitaled2/gp2_carriers/data/raw_genotypes/SAS/SAS_release9_vwb_snps.log.\n",
      "Options in effect:\n",
      "  --export Av\n",
      "  --extract /home/vitaled2/gp2_carriers/data/raw_genotypes/SAS/SAS_release9_vwb_temp_snps.txt\n",
      "  --freq\n",
      "  --out /home/vitaled2/gp2_carriers/data/raw_genotypes/SAS/SAS_release9_vwb_snps\n",
      "  --pfile /home/vitaled2/gp2_carriers/data/raw_genotypes/SAS/SAS_release9_vwb\n",
      "\n",
      "Start time: Fri Feb 14 18:49:12 2025\n",
      "32105 MiB RAM detected, ~30018 available; reserving 16052 MiB for main\n",
      "workspace.\n",
      "Using up to 8 compute threads.\n",
      "731 samples (262 females, 469 males; 731 founders) loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/SAS/SAS_release9_vwb.psam.\n",
      "1919616 variants loaded from\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/SAS/SAS_release9_vwb.pvar.\n",
      "1 binary phenotype loaded (235 cases, 217 controls).\n",
      "--extract: 24 variants remaining.\n",
      "Calculating allele frequencies... 252941done.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/SAS/SAS_release9_vwb_snps.afreq\n",
      ".\n",
      "24 variants remaining after main filters.\n",
      "--export Av to\n",
      "/home/vitaled2/gp2_carriers/data/raw_genotypes/SAS/SAS_release9_vwb_snps.traw\n",
      "... 121620252933374145505458626670757983879195done.\n",
      "End time: Fri Feb 14 18:49:13 2025\n"
     ]
    }
   ],
   "source": [
    "from carriers_api.app.carriers import extract_carriers, combine_carrier_files\n",
    "\n",
    "output_dir = f'{data_dir}/api_test/outputs'\n",
    "\n",
    "results_by_label = {}\n",
    "for label in labels:\n",
    "    results = extract_carriers(\n",
    "        geno_path=f'{geno_dir}/{label}/{label}_release9_vwb',\n",
    "        snplist_path=f'{data_dir}/snps_out.csv',\n",
    "        out_path=f'{output_dir}/{label}',\n",
    "        return_dfs=True\n",
    "    )\n",
    "    results_by_label[label] = results\n",
    "\n",
    "combined_results = combine_carrier_files(\n",
    "    results_by_label=results_by_label,\n",
    "    key_file=f'{data_dir}/nba_app_key.csv',\n",
    "    output_dir=output_dir\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>hg19</th>\n",
       "      <th>hg38</th>\n",
       "      <th>ancestry</th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>(C)M</th>\n",
       "      <th>POS</th>\n",
       "      <th>COUNTED</th>\n",
       "      <th>...</th>\n",
       "      <th>ALT_FREQS_EAS</th>\n",
       "      <th>OBS_CT_EAS</th>\n",
       "      <th>ALT_FREQS_EUR</th>\n",
       "      <th>OBS_CT_EUR</th>\n",
       "      <th>ALT_FREQS_FIN</th>\n",
       "      <th>OBS_CT_FIN</th>\n",
       "      <th>ALT_FREQS_MDE</th>\n",
       "      <th>OBS_CT_MDE</th>\n",
       "      <th>ALT_FREQS_SAS</th>\n",
       "      <th>OBS_CT_SAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:155204987-C-T</td>\n",
       "      <td>rs80356771</td>\n",
       "      <td>1:155204987:G:A</td>\n",
       "      <td>1:155235196:G:A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1:155204987-C-T</td>\n",
       "      <td>0</td>\n",
       "      <td>155235196</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001949</td>\n",
       "      <td>11800</td>\n",
       "      <td>0.000664</td>\n",
       "      <td>102480</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seq_rs80356771.1_ilmnrev_ilmnF2BT</td>\n",
       "      <td>rs80356771</td>\n",
       "      <td>1:155204987:G:A</td>\n",
       "      <td>1:155235196:G:A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Seq_rs80356771.1_ilmnrev_ilmnF2BT</td>\n",
       "      <td>0</td>\n",
       "      <td>155235196</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001949</td>\n",
       "      <td>11798</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>102394</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs80356771</td>\n",
       "      <td>rs80356771</td>\n",
       "      <td>1:155204987:G:A</td>\n",
       "      <td>1:155235196:G:A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>rs80356771</td>\n",
       "      <td>0</td>\n",
       "      <td>155235196</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001949</td>\n",
       "      <td>11798</td>\n",
       "      <td>0.000683</td>\n",
       "      <td>102492</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exm106220</td>\n",
       "      <td>rs75548401</td>\n",
       "      <td>1:155206037:G:A</td>\n",
       "      <td>1:155236246:G:A</td>\n",
       "      <td>AMISH/ EUR</td>\n",
       "      <td>1</td>\n",
       "      <td>exm106220</td>\n",
       "      <td>0</td>\n",
       "      <td>155236246</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000339</td>\n",
       "      <td>11798</td>\n",
       "      <td>0.012223</td>\n",
       "      <td>102434</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>272</td>\n",
       "      <td>0.000577</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs75548401</td>\n",
       "      <td>rs75548401</td>\n",
       "      <td>1:155206037:G:A</td>\n",
       "      <td>1:155236246:G:A</td>\n",
       "      <td>AMISH/ EUR</td>\n",
       "      <td>1</td>\n",
       "      <td>rs75548401</td>\n",
       "      <td>0</td>\n",
       "      <td>155236246</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>11788</td>\n",
       "      <td>0.012061</td>\n",
       "      <td>102226</td>\n",
       "      <td>0.033088</td>\n",
       "      <td>272</td>\n",
       "      <td>0.000579</td>\n",
       "      <td>1728</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>seq_rs75548401</td>\n",
       "      <td>rs75548401</td>\n",
       "      <td>1:155206037:G:A</td>\n",
       "      <td>1:155236246:G:A</td>\n",
       "      <td>AMISH/ EUR</td>\n",
       "      <td>1</td>\n",
       "      <td>seq_rs75548401</td>\n",
       "      <td>0</td>\n",
       "      <td>155236246</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>11804</td>\n",
       "      <td>0.012260</td>\n",
       "      <td>102446</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>272</td>\n",
       "      <td>0.000577</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Seq_rs34424986.1_ilmnrev_ilmnF2BT</td>\n",
       "      <td>rs34424986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:161785820:G:A</td>\n",
       "      <td>MULT</td>\n",
       "      <td>6</td>\n",
       "      <td>Seq_rs34424986.1_ilmnrev_ilmnF2BT</td>\n",
       "      <td>0</td>\n",
       "      <td>161785820</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11796</td>\n",
       "      <td>0.001060</td>\n",
       "      <td>101902</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>exm593505</td>\n",
       "      <td>rs34424986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:161785820:G:A</td>\n",
       "      <td>MULT</td>\n",
       "      <td>6</td>\n",
       "      <td>exm593505</td>\n",
       "      <td>0</td>\n",
       "      <td>161785820</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000170</td>\n",
       "      <td>11794</td>\n",
       "      <td>0.003641</td>\n",
       "      <td>102452</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs34424986</td>\n",
       "      <td>rs34424986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6:161785820:G:A</td>\n",
       "      <td>MULT</td>\n",
       "      <td>6</td>\n",
       "      <td>rs34424986</td>\n",
       "      <td>0</td>\n",
       "      <td>161785820</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>11798</td>\n",
       "      <td>0.003815</td>\n",
       "      <td>102480</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>12:40692927-A-G</td>\n",
       "      <td>rs34805604</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12:40299125:A:G</td>\n",
       "      <td>MULT</td>\n",
       "      <td>12</td>\n",
       "      <td>12:40692927-A-G</td>\n",
       "      <td>0</td>\n",
       "      <td>40299125</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11802</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>102486</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Seq_rs34805604</td>\n",
       "      <td>rs34805604</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12:40299125:A:G</td>\n",
       "      <td>MULT</td>\n",
       "      <td>12</td>\n",
       "      <td>Seq_rs34805604</td>\n",
       "      <td>0</td>\n",
       "      <td>40299125</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11804</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>102478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Seq_rs33939927.1_ilmnfwd_ilmnF2BT</td>\n",
       "      <td>rs33939927</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12:40310434:C:A</td>\n",
       "      <td>MULT</td>\n",
       "      <td>12</td>\n",
       "      <td>Seq_rs33939927.1_ilmnfwd_ilmnF2BT</td>\n",
       "      <td>0</td>\n",
       "      <td>40310434</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>11800</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>102474</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Seq_rs33939927.2_ilmnfwd_ilmnF2BT</td>\n",
       "      <td>rs33939927</td>\n",
       "      <td>12:40704236:C:G</td>\n",
       "      <td>12:40310434:C:G</td>\n",
       "      <td>MULT</td>\n",
       "      <td>12</td>\n",
       "      <td>Seq_rs33939927.2_ilmnfwd_ilmnF2BT</td>\n",
       "      <td>0</td>\n",
       "      <td>40310434</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11798</td>\n",
       "      <td>0.000927</td>\n",
       "      <td>102478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>12:40704237-G-A</td>\n",
       "      <td>rs34995376</td>\n",
       "      <td>12:40704237:G:A</td>\n",
       "      <td>12:40310435:G:A</td>\n",
       "      <td>MULT</td>\n",
       "      <td>12</td>\n",
       "      <td>12:40704237-G-A</td>\n",
       "      <td>0</td>\n",
       "      <td>40310435</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000169</td>\n",
       "      <td>11804</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>102476</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Seq_rs34995376</td>\n",
       "      <td>rs34995376</td>\n",
       "      <td>12:40704237:G:A</td>\n",
       "      <td>12:40310435:G:A</td>\n",
       "      <td>MULT</td>\n",
       "      <td>12</td>\n",
       "      <td>Seq_rs34995376</td>\n",
       "      <td>0</td>\n",
       "      <td>40310435</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000169</td>\n",
       "      <td>11802</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>102480</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>12:40714916-A-G</td>\n",
       "      <td>rs35801418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12:40321114:A:G</td>\n",
       "      <td>MULT</td>\n",
       "      <td>12</td>\n",
       "      <td>12:40714916-A-G</td>\n",
       "      <td>0</td>\n",
       "      <td>40321114</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11784</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>102302</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>272</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1730</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Seq_rs35801418</td>\n",
       "      <td>rs35801418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12:40321114:A:G</td>\n",
       "      <td>MULT</td>\n",
       "      <td>12</td>\n",
       "      <td>Seq_rs35801418</td>\n",
       "      <td>0</td>\n",
       "      <td>40321114</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11800</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>102434</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>272</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1730</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>exm994671</td>\n",
       "      <td>rs34637584</td>\n",
       "      <td>12:40734202:G:A</td>\n",
       "      <td>12:40340400:G:A</td>\n",
       "      <td>AJ/ EUR</td>\n",
       "      <td>12</td>\n",
       "      <td>exm994671</td>\n",
       "      <td>0</td>\n",
       "      <td>40340400</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>11802</td>\n",
       "      <td>0.004225</td>\n",
       "      <td>102484</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.039838</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>rs34637584</td>\n",
       "      <td>rs34637584</td>\n",
       "      <td>12:40734202:G:A</td>\n",
       "      <td>12:40340400:G:A</td>\n",
       "      <td>AJ/ EUR</td>\n",
       "      <td>12</td>\n",
       "      <td>rs34637584</td>\n",
       "      <td>0</td>\n",
       "      <td>40340400</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>11804</td>\n",
       "      <td>0.004225</td>\n",
       "      <td>102480</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.039838</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>seq_rs34637584</td>\n",
       "      <td>rs34637584</td>\n",
       "      <td>12:40734202:G:A</td>\n",
       "      <td>12:40340400:G:A</td>\n",
       "      <td>AJ/ EUR</td>\n",
       "      <td>12</td>\n",
       "      <td>seq_rs34637584</td>\n",
       "      <td>0</td>\n",
       "      <td>40340400</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>11802</td>\n",
       "      <td>0.004235</td>\n",
       "      <td>102486</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.039838</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>exm994721</td>\n",
       "      <td>rs34778348</td>\n",
       "      <td>12:40757328:G:A</td>\n",
       "      <td>12:40363526:G:A</td>\n",
       "      <td>EAS</td>\n",
       "      <td>12</td>\n",
       "      <td>exm994721</td>\n",
       "      <td>0</td>\n",
       "      <td>40363526</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.031271</td>\n",
       "      <td>11800</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>102474</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1730</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>rs34778348</td>\n",
       "      <td>rs34778348</td>\n",
       "      <td>12:40757328:G:A</td>\n",
       "      <td>12:40363526:G:A</td>\n",
       "      <td>EAS</td>\n",
       "      <td>12</td>\n",
       "      <td>rs34778348</td>\n",
       "      <td>0</td>\n",
       "      <td>40363526</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.031276</td>\n",
       "      <td>11798</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>102450</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1730</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>seq_rs34778348</td>\n",
       "      <td>rs34778348</td>\n",
       "      <td>12:40757328:G:A</td>\n",
       "      <td>12:40363526:G:A</td>\n",
       "      <td>EAS</td>\n",
       "      <td>12</td>\n",
       "      <td>seq_rs34778348</td>\n",
       "      <td>0</td>\n",
       "      <td>40363526</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.031276</td>\n",
       "      <td>11798</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>102400</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1728</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rs104893878</td>\n",
       "      <td>rs104893878</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4:89835580:C:G</td>\n",
       "      <td>EUR</td>\n",
       "      <td>4</td>\n",
       "      <td>rs104893878</td>\n",
       "      <td>0</td>\n",
       "      <td>89835580</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11804</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>102474</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>274</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24 rows Ã— 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   id         rsid             hg19  \\\n",
       "0                     1:155204987-C-T   rs80356771  1:155204987:G:A   \n",
       "1   Seq_rs80356771.1_ilmnrev_ilmnF2BT   rs80356771  1:155204987:G:A   \n",
       "2                          rs80356771   rs80356771  1:155204987:G:A   \n",
       "3                           exm106220   rs75548401  1:155206037:G:A   \n",
       "4                          rs75548401   rs75548401  1:155206037:G:A   \n",
       "5                      seq_rs75548401   rs75548401  1:155206037:G:A   \n",
       "6   Seq_rs34424986.1_ilmnrev_ilmnF2BT   rs34424986              NaN   \n",
       "7                           exm593505   rs34424986              NaN   \n",
       "8                          rs34424986   rs34424986              NaN   \n",
       "9                     12:40692927-A-G   rs34805604              NaN   \n",
       "10                     Seq_rs34805604   rs34805604              NaN   \n",
       "11  Seq_rs33939927.1_ilmnfwd_ilmnF2BT   rs33939927              NaN   \n",
       "12  Seq_rs33939927.2_ilmnfwd_ilmnF2BT   rs33939927  12:40704236:C:G   \n",
       "13                    12:40704237-G-A   rs34995376  12:40704237:G:A   \n",
       "14                     Seq_rs34995376   rs34995376  12:40704237:G:A   \n",
       "15                    12:40714916-A-G   rs35801418              NaN   \n",
       "16                     Seq_rs35801418   rs35801418              NaN   \n",
       "17                          exm994671   rs34637584  12:40734202:G:A   \n",
       "18                         rs34637584   rs34637584  12:40734202:G:A   \n",
       "19                     seq_rs34637584   rs34637584  12:40734202:G:A   \n",
       "20                          exm994721   rs34778348  12:40757328:G:A   \n",
       "21                         rs34778348   rs34778348  12:40757328:G:A   \n",
       "22                     seq_rs34778348   rs34778348  12:40757328:G:A   \n",
       "23                        rs104893878  rs104893878              NaN   \n",
       "\n",
       "               hg38    ancestry  CHR                                SNP  (C)M  \\\n",
       "0   1:155235196:G:A         NaN    1                    1:155204987-C-T     0   \n",
       "1   1:155235196:G:A         NaN    1  Seq_rs80356771.1_ilmnrev_ilmnF2BT     0   \n",
       "2   1:155235196:G:A         NaN    1                         rs80356771     0   \n",
       "3   1:155236246:G:A  AMISH/ EUR    1                          exm106220     0   \n",
       "4   1:155236246:G:A  AMISH/ EUR    1                         rs75548401     0   \n",
       "5   1:155236246:G:A  AMISH/ EUR    1                     seq_rs75548401     0   \n",
       "6   6:161785820:G:A        MULT    6  Seq_rs34424986.1_ilmnrev_ilmnF2BT     0   \n",
       "7   6:161785820:G:A        MULT    6                          exm593505     0   \n",
       "8   6:161785820:G:A        MULT    6                         rs34424986     0   \n",
       "9   12:40299125:A:G        MULT   12                    12:40692927-A-G     0   \n",
       "10  12:40299125:A:G        MULT   12                     Seq_rs34805604     0   \n",
       "11  12:40310434:C:A        MULT   12  Seq_rs33939927.1_ilmnfwd_ilmnF2BT     0   \n",
       "12  12:40310434:C:G        MULT   12  Seq_rs33939927.2_ilmnfwd_ilmnF2BT     0   \n",
       "13  12:40310435:G:A        MULT   12                    12:40704237-G-A     0   \n",
       "14  12:40310435:G:A        MULT   12                     Seq_rs34995376     0   \n",
       "15  12:40321114:A:G        MULT   12                    12:40714916-A-G     0   \n",
       "16  12:40321114:A:G        MULT   12                     Seq_rs35801418     0   \n",
       "17  12:40340400:G:A    AJ/ EUR    12                          exm994671     0   \n",
       "18  12:40340400:G:A    AJ/ EUR    12                         rs34637584     0   \n",
       "19  12:40340400:G:A    AJ/ EUR    12                     seq_rs34637584     0   \n",
       "20  12:40363526:G:A         EAS   12                          exm994721     0   \n",
       "21  12:40363526:G:A         EAS   12                         rs34778348     0   \n",
       "22  12:40363526:G:A         EAS   12                     seq_rs34778348     0   \n",
       "23   4:89835580:C:G         EUR    4                        rs104893878     0   \n",
       "\n",
       "          POS COUNTED  ... ALT_FREQS_EAS OBS_CT_EAS ALT_FREQS_EUR OBS_CT_EUR  \\\n",
       "0   155235196       G  ...      0.001949      11800      0.000664     102480   \n",
       "1   155235196       G  ...      0.001949      11798      0.000684     102394   \n",
       "2   155235196       G  ...      0.001949      11798      0.000683     102492   \n",
       "3   155236246       G  ...      0.000339      11798      0.012223     102434   \n",
       "4   155236246       G  ...      0.000254      11788      0.012061     102226   \n",
       "5   155236246       G  ...      0.000254      11804      0.012260     102446   \n",
       "6   161785820       G  ...      0.000000      11796      0.001060     101902   \n",
       "7   161785820       G  ...      0.000170      11794      0.003641     102452   \n",
       "8   161785820       G  ...      0.000085      11798      0.003815     102480   \n",
       "9    40299125       A  ...      0.000000      11802      0.000000     102486   \n",
       "10   40299125       A  ...      0.000000      11804      0.000010     102478   \n",
       "11   40310434       C  ...      0.000085      11800      0.000020     102474   \n",
       "12   40310434       C  ...      0.000000      11798      0.000927     102478   \n",
       "13   40310435       G  ...      0.000169      11804      0.000107     102476   \n",
       "14   40310435       G  ...      0.000169      11802      0.000107     102480   \n",
       "15   40321114       A  ...      0.000000      11784      0.000049     102302   \n",
       "16   40321114       A  ...      0.000000      11800      0.000029     102434   \n",
       "17   40340400       G  ...      0.000085      11802      0.004225     102484   \n",
       "18   40340400       G  ...      0.000085      11804      0.004225     102480   \n",
       "19   40340400       G  ...      0.000085      11802      0.004235     102486   \n",
       "20   40363526       G  ...      0.031271      11800      0.000068     102474   \n",
       "21   40363526       G  ...      0.031276      11798      0.000059     102450   \n",
       "22   40363526       G  ...      0.031276      11798      0.000078     102400   \n",
       "23   89835580       G  ...      0.000000      11804      0.000000     102474   \n",
       "\n",
       "    ALT_FREQS_FIN OBS_CT_FIN ALT_FREQS_MDE OBS_CT_MDE  ALT_FREQS_SAS  \\\n",
       "0        0.000000        274      0.000000       1732       0.000000   \n",
       "1        0.000000        274      0.000000       1732       0.000000   \n",
       "2        0.000000        274      0.000000       1732       0.000000   \n",
       "3        0.029412        272      0.000577       1732       0.000684   \n",
       "4        0.033088        272      0.000579       1728       0.000685   \n",
       "5        0.029412        272      0.000577       1732       0.000684   \n",
       "6        0.000000        274      0.000000       1732       0.000000   \n",
       "7        0.000000        274      0.000000       1732       0.000684   \n",
       "8        0.000000        274      0.000000       1732       0.000684   \n",
       "9        0.000000        274      0.000000       1732       0.000000   \n",
       "10       0.000000        274      0.000000       1732       0.000000   \n",
       "11       0.000000        274      0.000000       1732       0.000000   \n",
       "12       0.000000        274      0.000000       1732       0.000000   \n",
       "13       0.000000        274      0.000000       1732       0.000000   \n",
       "14       0.000000        274      0.000000       1732       0.000000   \n",
       "15       0.000000        272      0.000000       1730       0.000000   \n",
       "16       0.000000        272      0.000000       1730       0.000000   \n",
       "17       0.000000        274      0.039838       1732       0.000000   \n",
       "18       0.000000        274      0.039838       1732       0.000000   \n",
       "19       0.000000        274      0.039838       1732       0.000000   \n",
       "20       0.000000        274      0.000000       1730       0.000000   \n",
       "21       0.000000        274      0.000000       1730       0.000000   \n",
       "22       0.000000        274      0.000000       1728       0.000000   \n",
       "23       0.000000        274      0.000000       1732       0.000000   \n",
       "\n",
       "    OBS_CT_SAS  \n",
       "0         1462  \n",
       "1         1460  \n",
       "2         1462  \n",
       "3         1462  \n",
       "4         1460  \n",
       "5         1462  \n",
       "6         1460  \n",
       "7         1462  \n",
       "8         1462  \n",
       "9         1462  \n",
       "10        1462  \n",
       "11        1462  \n",
       "12        1462  \n",
       "13        1462  \n",
       "14        1462  \n",
       "15        1460  \n",
       "16        1462  \n",
       "17        1462  \n",
       "18        1462  \n",
       "19        1462  \n",
       "20        1462  \n",
       "21        1462  \n",
       "22        1460  \n",
       "23        1462  \n",
       "\n",
       "[24 rows x 40 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "carriers_string = pd.read_csv(combined_results['carriers_string'])\n",
    "var_info = pd.read_csv(combined_results['var_info'])\n",
    "var_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IID</th>\n",
       "      <th>study</th>\n",
       "      <th>ancestry</th>\n",
       "      <th>GBA1_p.R502C_1:155204987-C-T</th>\n",
       "      <th>GBA1_p.R502C_Seq_rs80356771.1_ilmnrev_ilmnF2BT</th>\n",
       "      <th>GBA1_p.R502C_rs80356771</th>\n",
       "      <th>GBA1_p.T408M_exm106220</th>\n",
       "      <th>GBA1_p.T408M_rs75548401</th>\n",
       "      <th>GBA1_p.T408M_seq_rs75548401</th>\n",
       "      <th>PRKN_R275W_Seq_rs34424986.1_ilmnrev_ilmnF2BT</th>\n",
       "      <th>...</th>\n",
       "      <th>LRRK2_R1441H_Seq_rs34995376</th>\n",
       "      <th>LRRK2_Y1699C_12:40714916-A-G</th>\n",
       "      <th>LRRK2_Y1699C_Seq_rs35801418</th>\n",
       "      <th>LRRK2_G2019S_exm994671</th>\n",
       "      <th>LRRK2_G2019S_rs34637584</th>\n",
       "      <th>LRRK2_G2019S_seq_rs34637584</th>\n",
       "      <th>LRRK2_G2385D_exm994721</th>\n",
       "      <th>LRRK2_G2385D_rs34778348</th>\n",
       "      <th>LRRK2_G2385D_seq_rs34778348</th>\n",
       "      <th>SNCA_A30P_rs104893878</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>BLAACPD-KPM_000032</td>\n",
       "      <td>BLAACPD-KPM</td>\n",
       "      <td>AAC</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1412</th>\n",
       "      <td>BLAACPD-UFL_000002</td>\n",
       "      <td>BLAACPD-UFL</td>\n",
       "      <td>AFR</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5121</th>\n",
       "      <td>LCC_000008</td>\n",
       "      <td>LCC</td>\n",
       "      <td>AJ</td>\n",
       "      <td>./.</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/G2019S</td>\n",
       "      <td>WT/G2019S</td>\n",
       "      <td>WT/G2019S</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9269</th>\n",
       "      <td>LARGEPD_001828</td>\n",
       "      <td>LARGEPD</td>\n",
       "      <td>AMR</td>\n",
       "      <td>./.</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10689</th>\n",
       "      <td>PDGNRTN_007257</td>\n",
       "      <td>PDGNRTN</td>\n",
       "      <td>AMR</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69623</th>\n",
       "      <td>VIPD_000214</td>\n",
       "      <td>VIPD</td>\n",
       "      <td>EUR</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69760</th>\n",
       "      <td>VIPD_000369</td>\n",
       "      <td>VIPD</td>\n",
       "      <td>EUR</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69881</th>\n",
       "      <td>VIPD_000512</td>\n",
       "      <td>VIPD</td>\n",
       "      <td>EUR</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70045</th>\n",
       "      <td>VIPD_000741</td>\n",
       "      <td>VIPD</td>\n",
       "      <td>EUR</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70095</th>\n",
       "      <td>YMS_000047</td>\n",
       "      <td>YMS</td>\n",
       "      <td>EUR</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/p.R502C</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>...</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "      <td>WT/WT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      IID        study ancestry GBA1_p.R502C_1:155204987-C-T  \\\n",
       "32     BLAACPD-KPM_000032  BLAACPD-KPM      AAC                   WT/p.R502C   \n",
       "1412   BLAACPD-UFL_000002  BLAACPD-UFL      AFR                   WT/p.R502C   \n",
       "5121           LCC_000008          LCC       AJ                          ./.   \n",
       "9269       LARGEPD_001828      LARGEPD      AMR                          ./.   \n",
       "10689      PDGNRTN_007257      PDGNRTN      AMR                   WT/p.R502C   \n",
       "...                   ...          ...      ...                          ...   \n",
       "69623         VIPD_000214         VIPD      EUR                   WT/p.R502C   \n",
       "69760         VIPD_000369         VIPD      EUR                   WT/p.R502C   \n",
       "69881         VIPD_000512         VIPD      EUR                   WT/p.R502C   \n",
       "70045         VIPD_000741         VIPD      EUR                   WT/p.R502C   \n",
       "70095          YMS_000047          YMS      EUR                   WT/p.R502C   \n",
       "\n",
       "      GBA1_p.R502C_Seq_rs80356771.1_ilmnrev_ilmnF2BT GBA1_p.R502C_rs80356771  \\\n",
       "32                                        WT/p.R502C              WT/p.R502C   \n",
       "1412                                      WT/p.R502C              WT/p.R502C   \n",
       "5121                                           WT/WT                   WT/WT   \n",
       "9269                                           WT/WT                   WT/WT   \n",
       "10689                                     WT/p.R502C              WT/p.R502C   \n",
       "...                                              ...                     ...   \n",
       "69623                                     WT/p.R502C              WT/p.R502C   \n",
       "69760                                     WT/p.R502C              WT/p.R502C   \n",
       "69881                                     WT/p.R502C              WT/p.R502C   \n",
       "70045                                     WT/p.R502C              WT/p.R502C   \n",
       "70095                                     WT/p.R502C              WT/p.R502C   \n",
       "\n",
       "      GBA1_p.T408M_exm106220 GBA1_p.T408M_rs75548401  \\\n",
       "32                     WT/WT                   WT/WT   \n",
       "1412                   WT/WT                   WT/WT   \n",
       "5121                   WT/WT                   WT/WT   \n",
       "9269                   WT/WT                   WT/WT   \n",
       "10689                  WT/WT                   WT/WT   \n",
       "...                      ...                     ...   \n",
       "69623                  WT/WT                   WT/WT   \n",
       "69760                  WT/WT                   WT/WT   \n",
       "69881                  WT/WT                   WT/WT   \n",
       "70045                  WT/WT                   WT/WT   \n",
       "70095                  WT/WT                   WT/WT   \n",
       "\n",
       "      GBA1_p.T408M_seq_rs75548401  \\\n",
       "32                          WT/WT   \n",
       "1412                        WT/WT   \n",
       "5121                        WT/WT   \n",
       "9269                        WT/WT   \n",
       "10689                       WT/WT   \n",
       "...                           ...   \n",
       "69623                       WT/WT   \n",
       "69760                       WT/WT   \n",
       "69881                       WT/WT   \n",
       "70045                       WT/WT   \n",
       "70095                       WT/WT   \n",
       "\n",
       "      PRKN_R275W_Seq_rs34424986.1_ilmnrev_ilmnF2BT  ...  \\\n",
       "32                                           WT/WT  ...   \n",
       "1412                                         WT/WT  ...   \n",
       "5121                                         WT/WT  ...   \n",
       "9269                                         WT/WT  ...   \n",
       "10689                                        WT/WT  ...   \n",
       "...                                            ...  ...   \n",
       "69623                                        WT/WT  ...   \n",
       "69760                                        WT/WT  ...   \n",
       "69881                                        WT/WT  ...   \n",
       "70045                                        WT/WT  ...   \n",
       "70095                                        WT/WT  ...   \n",
       "\n",
       "      LRRK2_R1441H_Seq_rs34995376 LRRK2_Y1699C_12:40714916-A-G  \\\n",
       "32                          WT/WT                        WT/WT   \n",
       "1412                        WT/WT                        WT/WT   \n",
       "5121                        WT/WT                        WT/WT   \n",
       "9269                        WT/WT                        WT/WT   \n",
       "10689                       WT/WT                        WT/WT   \n",
       "...                           ...                          ...   \n",
       "69623                       WT/WT                        WT/WT   \n",
       "69760                       WT/WT                        WT/WT   \n",
       "69881                       WT/WT                        WT/WT   \n",
       "70045                       WT/WT                        WT/WT   \n",
       "70095                       WT/WT                        WT/WT   \n",
       "\n",
       "      LRRK2_Y1699C_Seq_rs35801418 LRRK2_G2019S_exm994671  \\\n",
       "32                          WT/WT                  WT/WT   \n",
       "1412                        WT/WT                  WT/WT   \n",
       "5121                        WT/WT              WT/G2019S   \n",
       "9269                        WT/WT                  WT/WT   \n",
       "10689                       WT/WT                  WT/WT   \n",
       "...                           ...                    ...   \n",
       "69623                       WT/WT                  WT/WT   \n",
       "69760                       WT/WT                  WT/WT   \n",
       "69881                       WT/WT                  WT/WT   \n",
       "70045                       WT/WT                  WT/WT   \n",
       "70095                       WT/WT                  WT/WT   \n",
       "\n",
       "      LRRK2_G2019S_rs34637584 LRRK2_G2019S_seq_rs34637584  \\\n",
       "32                      WT/WT                       WT/WT   \n",
       "1412                    WT/WT                       WT/WT   \n",
       "5121                WT/G2019S                   WT/G2019S   \n",
       "9269                    WT/WT                       WT/WT   \n",
       "10689                   WT/WT                       WT/WT   \n",
       "...                       ...                         ...   \n",
       "69623                   WT/WT                       WT/WT   \n",
       "69760                   WT/WT                       WT/WT   \n",
       "69881                   WT/WT                       WT/WT   \n",
       "70045                   WT/WT                       WT/WT   \n",
       "70095                   WT/WT                       WT/WT   \n",
       "\n",
       "      LRRK2_G2385D_exm994721 LRRK2_G2385D_rs34778348  \\\n",
       "32                     WT/WT                   WT/WT   \n",
       "1412                   WT/WT                   WT/WT   \n",
       "5121                   WT/WT                   WT/WT   \n",
       "9269                   WT/WT                   WT/WT   \n",
       "10689                  WT/WT                   WT/WT   \n",
       "...                      ...                     ...   \n",
       "69623                  WT/WT                   WT/WT   \n",
       "69760                  WT/WT                   WT/WT   \n",
       "69881                  WT/WT                   WT/WT   \n",
       "70045                  WT/WT                   WT/WT   \n",
       "70095                  WT/WT                   WT/WT   \n",
       "\n",
       "      LRRK2_G2385D_seq_rs34778348 SNCA_A30P_rs104893878  \n",
       "32                          WT/WT                 WT/WT  \n",
       "1412                        WT/WT                 WT/WT  \n",
       "5121                        WT/WT                 WT/WT  \n",
       "9269                        WT/WT                 WT/WT  \n",
       "10689                       WT/WT                 WT/WT  \n",
       "...                           ...                   ...  \n",
       "69623                       WT/WT                 WT/WT  \n",
       "69760                       WT/WT                 WT/WT  \n",
       "69881                       WT/WT                 WT/WT  \n",
       "70045                       WT/WT                 WT/WT  \n",
       "70095                       WT/WT                 WT/WT  \n",
       "\n",
       "[106 rows x 27 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "carriers_string[carriers_string['GBA1_p.R502C_1:155204987-C-T']!='WT/WT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All checked genotypes match!\n"
     ]
    }
   ],
   "source": [
    "def verify_genotype(raw_value, string_value, snp_name):\n",
    "    if pd.isna(raw_value):\n",
    "        return string_value == './.'\n",
    "    raw_value = int(raw_value)\n",
    "    if raw_value == 2:\n",
    "        return string_value == \"WT/WT\"\n",
    "    elif raw_value == 1:\n",
    "        return string_value.startswith(\"WT/\")\n",
    "    elif raw_value == 0:\n",
    "        return string_value.count(snp_name.split('_')[1]) == 2\n",
    "    else:\n",
    "        return string_value == \":/:\"\n",
    "\n",
    "def validate_carrier_data(traw_dir, combined_file, snp_info_file, samples_to_check=None):\n",
    "    \"\"\"\n",
    "    Validate that combined carrier data matches original traw files.\n",
    "    \n",
    "    Args:\n",
    "        traw_dir: Directory containing traw files\n",
    "        combined_file: Path to combined carriers file\n",
    "        snp_info_file: Path to SNP info file\n",
    "        samples_to_check: List of sample IDs to check (None for all)\n",
    "    \"\"\"\n",
    "    carriers = pd.read_csv(combined_file)\n",
    "    snp_df = pd.read_csv(snp_info_file)\n",
    "    \n",
    "    if samples_to_check is None:\n",
    "        samples_to_check = carriers['IID'].unique()[:5]  # Check first 5 samples by default\n",
    "    \n",
    "    mismatches = []\n",
    "    for sample_id in samples_to_check:\n",
    "        # Get ancestry for this sample\n",
    "        ancestry = carriers.loc[carriers['IID'] == sample_id, 'ancestry'].iloc[0]\n",
    "        \n",
    "        # Read corresponding traw file\n",
    "        traw_path = f'{traw_dir}/{ancestry}/{ancestry}_release9_vwb_snps.traw'\n",
    "        traw = pd.read_csv(traw_path, sep='\\t')\n",
    "        traw_merged = snp_df.merge(traw, how='left', left_on='id', right_on='SNP')\n",
    "        \n",
    "        # Check each SNP\n",
    "        for snp in snp_df['snp_name_full']:\n",
    "            raw_value = traw_merged.loc[\n",
    "                traw_merged['snp_name_full'] == snp, \n",
    "                f'0_{sample_id}'\n",
    "            ].iloc[0]\n",
    "            \n",
    "            combined_value = carriers.loc[\n",
    "                carriers['IID'] == sample_id,\n",
    "                snp\n",
    "            ].iloc[0]\n",
    "            \n",
    "            if not verify_genotype(raw_value, combined_value, snp):\n",
    "                mismatches.append({\n",
    "                    'sample': sample_id,\n",
    "                    'ancestry': ancestry,\n",
    "                    'snp': snp,\n",
    "                    'raw': raw_value,\n",
    "                    'combined': combined_value\n",
    "                })\n",
    "    \n",
    "    if mismatches:\n",
    "        print(\"Found mismatches:\")\n",
    "        print(pd.DataFrame(mismatches))\n",
    "    else:\n",
    "        print(\"All checked genotypes match!\")\n",
    "\n",
    "# Use it like this:\n",
    "validate_carrier_data(\n",
    "    traw_dir=geno_dir,\n",
    "    combined_file=f'{output_dir}/carriers_string_full.csv',\n",
    "    snp_info_file=f'{data_dir}/snps_out.csv',\n",
    "    samples_to_check=['BLAACPD-KPM_000032']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# from carriers_api.app.carriers import extract_carriers\n",
    "\n",
    "# output_dir = f'{data_dir}/api_test/outputs'\n",
    "# carriers_string_full_path = f'{output_dir}/carriers_string_full.csv'\n",
    "# carriers_int_full_path = f'{output_dir}/carriers_int_full.csv'\n",
    "# var_info_full_path = f'{output_dir}/var_info_full.csv'\n",
    "\n",
    "# carriers_string_full = pd.DataFrame()\n",
    "# carriers_int_full = pd.DataFrame()\n",
    "# label_var_info_full = pd.DataFrame()\n",
    "\n",
    "# for label in labels:\n",
    "#     results = extract_carriers(\n",
    "#         geno_path=f'{geno_dir}/{label}/{label}_release9_vwb',\n",
    "#         snplist_path=f'{data_dir}/snps_out.csv',\n",
    "#         out_path=f'{output_dir}/{label}'\n",
    "#     )\n",
    "\n",
    "#     label_carriers_string = results['carriers_string']\n",
    "#     label_carriers_int = results['carriers_int']\n",
    "#     label_var_info = results['var_info']\n",
    "\n",
    "#     label_var_info_full = pd.concat([label_var_info_full, label_var_info], ignore_index=True)\n",
    "\n",
    "#     carriers_string = pd.read_csv(label_carriers_string)\n",
    "#     carriers_string['IID'] = carriers_string['IID'].str.replace('0_', '')\n",
    "#     carriers_string.loc[:,'ancestry'] = label\n",
    "#     carriers_string_full = pd.concat([carriers_string_full, carriers_string], ignore_index=True)\n",
    "    \n",
    "#     carriers_int = pd.read_csv(label_carriers_int)\n",
    "#     carriers_int['IID'] = carriers_int['IID'].str.replace('0_', '')\n",
    "#     carriers_int.loc[:,'ancestry'] = label\n",
    "#     carriers_int_full = pd.concat([carriers_int_full, carriers_int], ignore_index=True)\n",
    "\n",
    "# variant_columns = [x for x in carriers_string_full.columns if x not in ['IID','ancestry']]\n",
    "# carriers_string_full_out = carriers_string_full[['IID', 'ancestry'] + variant_columns]\n",
    "# carriers_string_full_out[variant_columns] = carriers_string_full_out[variant_columns].fillna('./.')\n",
    "# carriers_string_full_out_merge = carriers_string_full_out.merge(key[['IID','study']], how='left', on='IID')\n",
    "# carriers_string_final = carriers_string_full_out_merge[['IID', 'study', 'ancestry'] + variant_columns]\n",
    "# carriers_string_final.to_csv(carriers_string_full_path, index=False)\n",
    "\n",
    "# # Process and save final integer format\n",
    "# carriers_int_full_out = carriers_int_full[['IID', 'ancestry'] + [x for x in carriers_int_full.columns if x not in ['IID','ancestry']]]\n",
    "# carriers_int_full_out_merge = carriers_int_full_out.merge(key[['IID','study']], how='left', on='IID')\n",
    "# carriers_int_final = carriers_int_full_out_merge[['IID', 'study', 'ancestry'] + variant_columns]\n",
    "# carriers_int_final.to_csv(carriers_int_full_path, index=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# carriers_string_full_path = f'{data_dir}/api_test/outputs/carriers_string_full.csv'\n",
    "# carriers_int_full_path = f'{data_dir}/api_test/outputs/carriers_int_full.csv'\n",
    "\n",
    "# carriers_string_full = pd.DataFrame()\n",
    "# carriers_int_full = pd.DataFrame()\n",
    "\n",
    "# for label in labels:\n",
    "#     label_carriers_string = f'{data_dir}/api_test/outputs/{label}_carriers_string.csv'\n",
    "#     label_carriers_int = f'{data_dir}/api_test/outputs/{label}_carriers_int.csv'\n",
    "    \n",
    "#     carriers_string = pd.read_csv(label_carriers_string)\n",
    "#     carriers_string['IID'] = carriers_string['IID'].str.replace('0_', '')\n",
    "#     carriers_string.loc[:,'ancestry'] = label\n",
    "#     carriers_string_full = pd.concat([carriers_string_full, carriers_string], ignore_index=True)\n",
    "    \n",
    "#     carriers_int = pd.read_csv(label_carriers_int)\n",
    "#     carriers_int['IID'] = carriers_int['IID'].str.replace('0_', '')\n",
    "#     carriers_int.loc[:,'ancestry'] = label\n",
    "#     carriers_int_full = pd.concat([carriers_int_full, carriers_int], ignore_index=True)\n",
    "\n",
    "# # Process and save final string format\n",
    "# variant_columns = [x for x in carriers_string_full.columns if x not in ['IID','ancestry']]\n",
    "# carriers_string_full_out = carriers_string_full[['IID', 'ancestry'] + variant_columns]\n",
    "# carriers_string_full_out[variant_columns] = carriers_string_full_out[variant_columns].fillna('./.')\n",
    "# carriers_string_full_out_merge = carriers_string_full_out.merge(key[['IID','study']], how='left', on='IID')\n",
    "# carriers_string_final = carriers_string_full_out_merge[['IID', 'study', 'ancestry'] + variant_columns]\n",
    "# carriers_string_final.to_csv(carriers_string_full_path, index=False)\n",
    "\n",
    "# # Process and save final integer format\n",
    "# carriers_int_full_out = carriers_int_full[['IID', 'ancestry'] + [x for x in carriers_int_full.columns if x not in ['IID','ancestry']]]\n",
    "# carriers_int_full_out_merge = carriers_int_full_out.merge(key[['IID','study']], how='left', on='IID')\n",
    "# carriers_int_final = carriers_int_full_out_merge[['IID', 'study', 'ancestry'] + variant_columns]\n",
    "# carriers_int_final.to_csv(carriers_int_full_path, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replaced by modules in carriers.py\n",
    "\n",
    "# for label in labels:\n",
    "#     label_geno_path = f'{geno_dir}/{label}/{label}_release9_vwb'\n",
    "#     label_snps_traw = f'{label_geno_path}_snps.traw'\n",
    "#     label_snps_afreq = f'{label_geno_path}_snps.afreq'\n",
    "#     label_carriers_var_info = f'{data_dir}/outputs/{label}_carriers_var_info.csv'\n",
    "#     label_carriers_string = f'{data_dir}/outputs/{label}_carriers_string.csv'\n",
    "#     label_carriers_int = f'{data_dir}/outputs/{label}_carriers_int.csv'\n",
    "\n",
    "#     snp_df = pd.read_csv(f'{data_dir}/snps_out.csv')\n",
    "#     snp_df['id'].to_csv(f'{data_dir}/{label}_snps.txt', header=False, index=False)\n",
    "\n",
    "#     extract_cmd = f\"plink2 --pfile {label_geno_path} --extract {data_dir}/{label}_snps.txt --export Av --freq --out {label_geno_path}_snps\"\n",
    "#     !{extract_cmd}\n",
    "    \n",
    "#     traw = pd.read_csv(label_snps_traw, sep='\\t')\n",
    "#     traw_merged = snp_df.merge(traw, how='left', left_on='id', right_on='SNP')\n",
    "    \n",
    "#     colnames = [\n",
    "#         'id',\n",
    "#         'rsid',\n",
    "#         'hg19',\n",
    "#         'hg38',\n",
    "#         'ancestry', \n",
    "#         'CHR',\n",
    "#         'SNP', \n",
    "#         '(C)M', \n",
    "#         'POS',\n",
    "#         'COUNTED',\n",
    "#         'ALT',\n",
    "#         'variant',\n",
    "#         'snp_name', \n",
    "#         'locus', \n",
    "#         'snp_name_full'\n",
    "#         ]\n",
    "#     var_cols = [x for x in colnames if x not in ['snp_name_full']]\n",
    "#     sample_cols = list(traw_merged.drop(columns=colnames).columns)\n",
    "\n",
    "#     traw_final = traw_merged.loc[:, colnames + sample_cols]\n",
    "    \n",
    "#     traw_out = traw_final.copy()\n",
    "#     traw_out[sample_cols] = traw_out.apply(\n",
    "#         lambda row: [genotype_to_string(row[col], row['snp_name']) for col in sample_cols],\n",
    "#         axis=1,\n",
    "#         result_type='expand'\n",
    "#     )\n",
    "\n",
    "#     gt_df = traw_final.drop(columns=var_cols).set_index('snp_name_full').T.reset_index()\n",
    "#     gt_df.columns.name = None\n",
    "#     gt_df.rename(columns={'index':'IID'}, inplace=True)\n",
    "\n",
    "#     # Process and save frequency info\n",
    "#     freq = pd.read_csv(label_snps_afreq, sep='\\t')\n",
    "#     freq.rename(columns={'ID':'SNP'}, inplace=True)\n",
    "#     var_info_df = traw_final.loc[:,var_cols]\n",
    "#     var_info_df.merge(freq, how='left', on='SNP')\n",
    "#     var_info_df.to_csv(label_carriers_var_info, index=False)\n",
    "    \n",
    "#     # Process and save string format\n",
    "#     carriers_string = traw_out.drop(columns=var_cols).set_index('snp_name_full').T.reset_index()\n",
    "#     carriers_string.columns.name = None\n",
    "#     carriers_string = carriers_string.fillna('./.')\n",
    "#     carriers_string = carriers_string.astype(str)\n",
    "#     carriers_string.rename(columns={'index':'IID'}, inplace=True)\n",
    "#     carriers_string.to_csv(label_carriers_string, index=False)\n",
    "\n",
    "#     # Process and save integer format\n",
    "#     carriers_int = traw_final.drop(columns=var_cols).set_index('snp_name_full').T.reset_index()\n",
    "#     carriers_int.columns.name = None\n",
    "#     carriers_int.rename(columns={'index':'IID'}, inplace=True)\n",
    "#     carriers_int.to_csv(label_carriers_int, index=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !gcloud storage cp {data_dir}/outputs/carriers_string_full.csv gs://gp2_carriers/\n",
    "# !gcloud storage cp {data_dir}/outputs/carriers_int_full.csv gs://gp2_carriers/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
